### PR TITLE
feat: Workflow - approval node changes

### PR DESCRIPTION
**Pull request by sohel-rp**
# Description

These changes include,

- Updates in User permission configuration for setting an user as approver
- Updates in CD pipeline creation to configure manual approval & required count
- Updates in Single App & App Group trigger views to perform & handle different approval actions
- data-testid for automation testing

Fixes - [AB#2753](https://dev.azure.com/DevtronLabs/Devtron/_workitems/edit/2753)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

All tests have been performed manually & via automation tests by performing different actions.

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/ApplicationGroup/Constants.ts [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import { DeploymentNodeType } from '../app/details/triggerView/types'
+import { DeploymentNodeType } from '@devtron-labs/devtron-fe-common-lib'
```
Comment (vivek-devtron) [inline src/components/ApplicationGroup/Constants.ts:1]: Please remove unused import

File: src/components/ApplicationGroup/Details/TriggerView/BulkCDTrigger.tsx [hunk 2]
```diff
@@ -91,12 +97,17 @@ export default function BulkCDTrigger({
...
         }
-        const _materialListMap: Record<string, any[]> = {}
+        const _cdMaterialResponse: Record<string, CDMaterialResponseType> = {}
         Promise.allSettled(_CDMaterialPromiseList)
             .then((responses) => {
                 responses.forEach((response, index) => {
                     if (response.status === 'fulfilled') {
-                        _materialListMap[response.value['appId']] = response.value['materialList']
+                        _cdMaterialResponse[response.value['appId']] = {
+                            approvalUsers: response.value['approvalUsers'],
```
Comment (vivek-devtron) [inline src/components/ApplicationGroup/Details/TriggerView/BulkCDTrigger.tsx:106]: Are we not moving this out?

File: src/components/ApplicationGroup/Details/TriggerView/EnvTriggerView.tsx [hunk 3]
```diff
@@ -535,19 +547,30 @@ export default function EnvTriggerView({ filteredAppIds }: AppGroupDetailDefault
...
+            abortControllerRef.current.signal,
+            isApprovalNode,
+        )
             .then((data) => {
                 let _selectedNode
                 const _workflows = [...filteredWorkflows].map((workflow) => {
                     const nodes = workflow.nodes.map((node) => {
                         if (cdNodeId == node.id && node.type === nodeType) {
-                            node[MATERIAL_TYPE.inputMaterialList] = data
+                            node[MATERIAL_TYPE.inputMaterialList] = data.materials
```
Comment (vivek-devtron) [inline src/components/ApplicationGroup/Details/TriggerView/EnvTriggerView.tsx:569]: Same here

File: src/components/app/details/cdDetails/DeploymentDetailSteps.tsx [hunk 4]
```diff
@@ -91,6 +104,46 @@ export default function DeploymentDetailSteps({ deploymentStatus, deploymentAppT
         </div>
     ) : (
         <div className="dc__mxw-1000 min-w-800">
+            {userApprovalMetadata && (
```
Comment (vivek-devtron) [inline src/components/app/details/cdDetails/DeploymentDetailSteps.tsx:107]: this should be coming from Lib

File: src/components/app/details/cicdHistory/Artifacts.tsx [hunk 5]
```diff
@@ -166,13 +168,41 @@ const CIProgressView = (): JSX.Element => {
...
 
-const CIListItem = ({ type, children }: CIListItemType) => {
+export const CIListItem = ({ type, userApprovalMetadata, triggeredBy, children }: CIListItemType) => {
     return (
         <div className={`mb-16 ci-artifact ci-artifact--${type}`} data-testid="hover-on-report-artifact">
-            <div className="bcn-1 flex br-4">
-                <img src={type === 'artifact' ? docker : folder} className="icon-dim-24" />
-            </div>
-            {children}
+            {type === 'approved-artifact' ? (
```
Comment (vivek-devtron) [inline src/components/app/details/cicdHistory/Artifacts.tsx:174]: Same here

File: src/components/app/details/cicdHistory/History.components.tsx [hunk 6]
```diff
@@ -102,6 +113,33 @@ export const GitChanges = ({ gitTriggers, ciMaterials }: GitChangesType) => {
                     </div>
                 ) : null
             })}
+            {artifact && userApprovalMetadata && (
```
Comment (vivek-devtron) [inline src/components/app/details/cicdHistory/History.components.tsx:116]: same here

File: src/components/app/details/triggerView/workflow.service.ts [hunk 7]
```diff
@@ -192,8 +192,18 @@ export function processWorkflow(
 
                         const cdNode = cdPipelineToNode(cdPipeline, dimensions, branch.parentId)
                         wf.nodes.push(cdNode)
+
+                        if (cdPipeline.userApprovalConfig?.requiredCount > 0) {
```
Comment (vivek-devtron) [inline src/components/app/details/triggerView/workflow.service.ts:196]: same here

File: src/components/app/details/triggerView/workflow/Workflow.tsx [hunk 8]
```diff
@@ -246,16 +251,33 @@ export class Workflow extends Component<WorkflowProps> {
...
+    onClickNodeEdge = (nodeId: number) => {
+        this.context.onClickCDMaterial(nodeId, DeploymentNodeType.CD, true)
+    }
+
     renderEdgeList() {
-        return this.getEdges().map((edgeNode) => {
+        const edges = this.getEdges()
+        return edges.map((edgeNode) => {
+            if (ApprovalNodeEdge) {
+                return (
```
Comment (vivek-devtron) [inline src/components/app/details/triggerView/workflow/Workflow.tsx:262]: same here

File: src/components/app/service.ts [hunk 9]
```diff
@@ -237,17 +243,38 @@ export const getCIMaterialList = (params, abortSignal: AbortSignal) => {
...
+        if (!response.result) {
+            return {
+                approvalUsers: [],
+                materials: [],
+                userApprovalConfig: null,
+                requestedUserId: 0,
+            }
+        }
+
+        return {
```
Comment (vivek-devtron) [inline src/components/app/service.ts:267]: same here

File: src/components/common/ContentCard/ContentCard.tsx [hunk 10]
```diff
@@ -1,8 +1,9 @@
 import React from 'react'
 import { NavLink } from 'react-router-dom'
-import { ConditionalWrap, noop } from '../helpers/Helpers'
+import { ConditionalWrap } from '../helpers/Helpers'
 import { CardContentDirection, CardLinkIconPlacement, ContentCardProps } from './ContentCard.types'
 import './ContentCard.scss'
+import { noop } from '@devtron-labs/devtron-fe-common-lib'
```
Comment (vivek-devtron) [inline src/components/common/ContentCard/ContentCard.tsx:6]: we should import conditionalWrap also from lib

File: src/components/userGroups/UserGroup.tsx [hunk 11]
```diff
@@ -848,11 +861,37 @@ export const DirectPermission: React.FC<DirectPermissionRow> = ({
...
                 {React.cloneElement(children[1])}
             </components.ValueContainer>
         )
     }
 
+    const handleApproverChange = () => {
+        handleDirectPermissionChange(APPROVER_ACTION, { name: APPROVER_ACTION.label })
+    }
+
+    const RoleMenuList = (props) => {
```
Comment (vivek-devtron) [inline src/components/userGroups/UserGroup.tsx:874]: Same here

File: src/components/workflowEditor/Workflow.tsx [hunk 12]
```diff
@@ -369,16 +373,34 @@ export class Workflow extends Component<WorkflowProps, WorkflowState> {
...
 
+    onClickNodeEdge = (nodeId: number) => {
+        const ciPipeline = this.props.nodes.find((nd) => nd.type == WorkflowNodeType.CI)
+        this.props.history.push(`workflow/${this.props.id}/ci-pipeline/${+ciPipeline?.id}/cd-pipeline/${nodeId}`)
+    }
+
     renderEdgeList() {
+        const edges = this.getEdges()
         return this.getEdges().map((edgeNode) => {
+            if (ApprovalNodeEdge) {
```
Comment (vivek-devtron) [inline src/components/workflowEditor/Workflow.tsx:384]: Same here

File: src/components/app/details/triggerView/cdMaterial.tsx [hunk 13]
```diff
@@ -386,60 +505,68 @@ export class CDMaterial extends Component<CDMaterialProps, CDMaterialState> {
...
                 {!hideSelector && (
-                    <div className="material-history__select-text w-auto">
-                        {mat.vulnerable ? (
-                            <span className="material-history__scan-error">Security Issues Found</span>
-                        ) : mat.isSelected ? (
-                            <Check className="dc__align-right icon-dim-24" />
-                        ) : (
-                            'Select'
-                        )}
+                    <div className="material-history__select-text fs-13 w-auto dc__no-text-transform flex right cursor-default">
```
Comment (vivek-devtron) [inline src/components/app/details/triggerView/cdMaterial.tsx:508]: same here

File: src/components/app/details/triggerView/cdMaterial.tsx [hunk 14]
```diff
@@ -967,18 +1175,65 @@ export class CDMaterial extends Component<CDMaterialProps, CDMaterialState> {
                         <img alt="close" src={close} />
                     </button>
                 </div>
-                {this.renderTriggerBody()}
-                {this.renderTriggerModalCTA()}
+                {isApprovalConfigured &&
```
Comment (vivek-devtron) [inline src/components/app/details/triggerView/cdMaterial.tsx:1178]: Same here

File: src/components/cdPipeline/CDPipeline.tsx [hunk 15]
```diff
@@ -1385,8 +1410,36 @@ export default class CDPipeline extends Component<CDPipelineProps, CDPipelineSta
...
+    }
+
+    renderAdvancedCD() {
+        return (
+            <>
+                {this.renderPipelineNameInput()}
+                <div className="divider mt-12 mb-12" />
+                {this.renderPreStage()}
+                <div className="divider mt-12 mb-12" />
+                {ManualApproval && (
```
Comment (vivek-devtron) [inline src/components/cdPipeline/CDPipeline.tsx:1429]: same here

File: src/components/ApplicationGroup/Constants.ts (hunk 1 above)
Comment (sohel-rp) [reply src/components/ApplicationGroup/Constants.ts:1]: This is used in title & some text consts

File: src/components/app/details/cdDetails/DeploymentDetailSteps.tsx (hunk 4 above)
Comment (sohel-rp) [reply src/components/app/details/cdDetails/DeploymentDetailSteps.tsx:107]: Done

File: src/components/app/details/cicdHistory/Artifacts.tsx (hunk 5 above)
Comment (sohel-rp) [reply src/components/app/details/cicdHistory/Artifacts.tsx:174]: Done. I did extract the ApprovalInfoTippy but have moved the whole artifact part

File: src/components/app/details/cicdHistory/History.components.tsx (hunk 6 above)
Comment (sohel-rp) [reply src/components/app/details/cicdHistory/History.components.tsx:116]: I think we can keep this as internal part which was specific has been moved

File: src/components/app/details/triggerView/EmptyStateCdMaterial.tsx
[the remark this answers is not in the document]
Comment (sohel-rp) [reply src/components/app/details/triggerView/EmptyStateCdMaterial.tsx:1]: Yes... Changed with GenericEmptyState.

File: src/components/userGroups/UserGroup.tsx (hunk 11 above)
Comment (sohel-rp) [reply src/components/userGroups/UserGroup.tsx:874]: This is already moved to ApprovalNodeEdge, right?

File: src/components/app/details/triggerView/cdMaterial.tsx (hunk 14 above)
Comment (sohel-rp) [reply src/components/app/details/triggerView/cdMaterial.tsx:1178]: Kept this intentionally

File: src/components/cdPipeline/CDPipeline.tsx (hunk 15 above)
Comment (sohel-rp) [reply src/components/cdPipeline/CDPipeline.tsx:1429]: Already moved to `ManualApproval`

**Review by vivek-devtron — APPROVED** [review]

**Review by kripanshdevtron — APPROVED** [review]

**Pull request merged**